### PR TITLE
iso_extract: Reimplement using 7zip (not requiring root)

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
So one of the drawbacks of the original implementation is that it required root for mounting/unmount the ISO image.

This is now no longer needed as we (optionally) use 7zip for extracting files from the ISO.

This PR includes:
- works without mount-capabilities (or root-privileges) when 7zip is installed
- new integration tests (including ~MacOSX and~ FreeBSD)

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

iso_extract

##### ANSIBLE VERSION

v2.4

### CODE REVIEW

**Comment by dagwieers** [comment]
cc: @jhoekx @ribbons @jctanner You may be interested in this modification !

File: lib/ansible/modules/files/iso_extract.py [hunk 1]
```diff
@@ -69,63 +87,106 @@
...
+        module.fail_json(msg='Executable "%s" is not found on the system' % executable)
 
     if not os.path.exists(dest):
         module.fail_json(msg='Directory "%s" does not exist' % dest)
 
     if not os.path.exists(os.path.dirname(image)):
         module.fail_json(msg='ISO image "%s" does not exist' % image)
 
+    result['files'] = []
+    extract_files = list(files)
```
Comment (ribbons) [inline lib/ansible/modules/files/iso_extract.py]: I'm probably just having a brain-fail due to the virus I'm suffering from, but isn't `files` already a list?

**Review by ribbons — COMMENTED** [review]
This sounds good to me as it should stop these appearing in Dolphin each time I run one of my playbooks:
screenshot_20170523_195008

However, is adding a dependency on `7z` going to be an irritation to some users?

File: lib/ansible/modules/files/iso_extract.py (hunk 1 above)
Comment (dagwieers) [reply lib/ansible/modules/files/iso_extract.py]: Sure, but that's the python way to make a new copy of a list.
You can also do `extract_files = files[:]`, but that just looks weird.

Comment (dagwieers) [reply lib/ansible/modules/files/iso_extract.py]: The only reason for doing this, is to support `force: no` mode which avoids making checksums twice (e.g. with large files) in favor of doing just an existence check. It's a trade-off.

Comment (ribbons) [reply lib/ansible/modules/files/iso_extract.py]: Gotcha. As you were :smile:

**Comment by dagwieers** [comment]
shipit

**Comment by dagwieers** [comment]
bot_status

**Comment by dagwieers** [comment]
@ribbons We could fall back to the old behaviour if 7z was not found ?

**Comment by ribbons** [comment]
> @ribbons We could fall back to the old behaviour if 7z was not found ?

That sounds like the best option to me - if it fails due to not being root, the error message could suggest installing `7z`.

**Comment by dagwieers** [comment]
Well, being able to mount is related on the CAP_SYS_ADMIN capability on Linux (not just root). Might be different on other Unix systems even. So we cannot make assumptions there...

**Comment by dagwieers** [comment]
@jhoekx @ribbons Implemented as discussed, please test !

File: lib/ansible/modules/files/iso_extract.py [hunk 2]
```diff
@@ -23,18 +24,26 @@
...
 description:
-- This module mounts an iso image in a temporary directory and extracts
-  files from there to a given destination.
-version_added: "2.3"
+- This module has two possible ways of operation.
+- If 7zip is installed on the system, this module extracts files from an ISO
+  into a temporary directory and copies files to a given destination,
+  if needed.
+- If the user has mount-capabilities (CAP_SYS_ADMIN on Linux) this module
+  mounts the ISO image to a remporary location, and copites files to a given
```
Comment (ribbons) [inline lib/ansible/modules/files/iso_extract.py]: Couple of typos: remporary & copites

**Review by ribbons — COMMENTED** [review]
Have tested the updated version with the fallback and it worked really well.  Spotted a couple of tiny typos that you might want to fix, but apart from that it looks good to me.

**Comment by dagwieers** [comment]
@jctanner Can we get this merged ?

**Comment by mattclay** [comment]
It would be good to test this in CI now that it can be done without additional privileges. To do so we'll just need to install 7z as part of the integration test.

Of course we'll still want tests for the non-7z case, but those will need to be skipped if we don't have adequate privileges. Perhaps there's a command we can execute during the test to determine if we have enough permissions to proceed with the non-7z portion of the testing.

**Comment by dagwieers** [comment]
> To do so we'll just need to install 7z as part of the integration test.

That part took me about 3 hours, and I still cannot get it to work for MacOSX. (FreeBSD was also hell) This a complete waste of my time, I am not interested in how all these platforms are supposed to get software installed properly, but still to get a stupid modification to a module upstream I need to jump through these silly hoops ?

Homebrew output:
```
Error: Running Homebrew as root is extremely dangerous. As Homebrew does not
drop privileges on installation you are giving all build scripts full access
to your system. As a result of the macOS sandbox not handling the root user
correctly HOMEBREW_NO_SANDBOX has been set so the sandbox will not be used. If
we have not merged a pull request to add privilege dropping by November 1st
2016 running Homebrew as root will be disabled. No Homebrew maintainers plan
to work on this functionality.\nError: Cowardly refusing to 'sudo brew update'
You can use brew with sudo, but only if the brew executable is owned by root.
However, this is both not recommended and completely unsupported so do so at
your own risk."
```

Macports doesn't seem to find the package at all (even though the website lists it).

**Comment by dagwieers** [comment]
So despite the MacOSX stuff, I think this is ready to be merged. I do not plan to spend any more time on this one. But I would like this to be released sooner rather than later... @jctanner @mattclay @jhoekx @ribbons 

Please review and sign-off with no delay ;-)

**Comment by ribbons** [comment]
shipit

**Comment by mattclay** [comment]
I've updated the tests to work on OS X: https://github.com/ansible/ansible/pull/27786

**Pull request merged**